server/src/main/java/org/elasticsearch/common/compress/DeflateCompressor.java

```diff
@@ -88,56 +142,64 @@ public StreamInput streamInput(StreamInput in) throws IOException {
             throw new IllegalArgumentException("Input stream is not compressed with DEFLATE!");
         }
 
-        final boolean nowrap = true;
-        final Inflater inflater = new Inflater(nowrap);
-        InputStream decompressedIn = new InflaterInputStream(in, inflater, BUFFER_SIZE);
-        decompressedIn = new BufferedInputStream(decompressedIn, BUFFER_SIZE);
-        return new InputStreamStreamInput(decompressedIn) {
-            final AtomicBoolean closed = new AtomicBoolean(false);
-
+        final ReleasableReference<Inflater> current = inflaterForStreamRef.get();
+        final Releasable releasable;
+        final Inflater inflater;
+        if (current.inUse) {
+            // Nested de-compression streams should not happen but we still handle them safely by using a fresh Inflater
+            inflater = new Inflater(true);
+            releasable = inflater::end;
+        } else {
+            inflater = current.get();
+            releasable = current;
+        }
+        InputStream decompressedIn = new InflaterInputStream(in, inflater, BUFFER_SIZE) {
+            @Override
             public void close() throws IOException {
                 try {
                     super.close();
                 } finally {
-                    if (closed.compareAndSet(false, true)) {
-                        // important to release native memory
-                        inflater.end();
-                    }
+                    releasable.close();
```

Comment:
No need for this kind of extra atomic boolean+check, the `BufferedInputStream` will only close once.

Reply:
Let's add this as a code comment